t/23-baseclass.t

```diff
@@ -617,4 +617,106 @@ subtest 'bouncer functions' => sub {
+    my $fail_res = $baseclass->verify_image({imgpath => "$Bin/imgsearch/kde-logo.png", mustmatch => 0});
+    is_deeply $fail_res, {candidates => []}, 'image not found (no candidates)' or diag explain $fail_res;
+
+    my $fake_image = Test::MockObject->new->mock(search => sub ($self, $needles, $threshold, $search_ratio) { (1, [qw(foo bar)]) });
+    my $tinycv_mock = Test::MockModule->new('tinycv')->redefine(read => $fake_image);
+    my $ok_res = $baseclass->verify_image({imgpath => "$Bin/imgsearch/kde-logo.png", mustmatch => 0});
+    is_deeply $ok_res, {found => 1, candidates => [qw(foo bar)]}, 'image found (mocked search)' or diag explain $ok_res;
+};
+
+subtest 'retrying assert screen' => sub {
+    my $needles_reloaded = 0;
+    my $mock = Test::MockModule->new('backend::baseclass')->redefine(reload_needles => sub ($self) { $needles_reloaded = 1 });
+    $baseclass->assert_screen_deadline(0);
+    combined_like {
+        $baseclass->retry_assert_screen({reload_needles => 1, timeout => 42})
+    } qr/cont_vm.*set_tags_to_assert: NO matching needles for foo/s, 'cont_vm called, set_tags_to_assert invoked';
+    ok $needles_reloaded, 'needles have been reloaded';
+    ok $baseclass->assert_screen_deadline, 'assert screen timeout set';
+};
+
+local $SIG{__DIE__} = 'DEFAULT';
+
+subtest 'error handling when checking socket' => sub {
+    my $rpc_mock = Test::MockModule->new('myjsonrpc')->redefine(read_json => {invalid => 'response'});
+    $baseclass->{cmdpipe} = 42;
+    throws_ok { $baseclass->check_socket(42) } qr/no command in.*invalid.*response/s, 'dies on invalid response';
+};
+
+subtest 'special cases of set_tags_to_assert' => sub {
+    combined_like { needle::init("$Bin/data") } qr/loaded \d+ needles/, 'needles loaded';
+
+    subtest 'invalid tags passed' => sub {
+        my $warning = warning {
+            combined_like {
+                my $res = $baseclass->set_tags_to_assert({mustmatch => [{invalid => 'tags'}]});
+                is_deeply $res, {tags => []}, 'empty set of tags returned for invalid needle' or diag explain $res;
+            } qr/NO matching needles for/, 'no match logged for invalid needle'
+        };
+        like $warning, qr/invalid needle passed <HASH>.*invalid.*tags/s, 'warning about invalid needle';
+        is_deeply $baseclass->assert_screen_needles, [], 'no assert screen needles assigned';
+    };
+
+    subtest 'multiple tags specified, multiple needles set for assertion' => sub {
+        my @tags = (qw(inst-welcome not-existing));
+        my $res = $baseclass->set_tags_to_assert({mustmatch => \@tags});
+        is_deeply $res, {tags => \@tags}, 'tags returned' or diag explain $res;
+        my $needles = $baseclass->assert_screen_needles;
+        is scalar @$needles, 2, 'matching needles assigned';
+        is $needles->[0]->{name}, 'welcome.ref', 'needle welcome.ref matched';
+        is $needles->[1]->{name}, 'inst-welcome-20140902', 'needle inst-welcome-20140902 matched';
```

Comment:
7:         #   Failed test 'needle welcome.ref matched'
7:         #   at ./t/23-baseclass.t line 702.
7:         #          got: 'inst-welcome-20140902'
7:         #     expected: 'welcome.ref'
7: 
7:         #   Failed test 'needle inst-welcome-20140902 matched'
7:         #   at ./t/23-baseclass.t line 703.
7:         #          got: 'welcome.ref'
7:         #     expected: 'inst-welcome-20140902'
7:         # Looks like you failed 2 tests of 4.
7: 
7:     #   Failed test 'multiple tags specified, multiple needles set for assertion'
7:     #   at ./t/23-baseclass.t line 704.
7:     # Looks like you failed 1 test of 3.

Reply:
I wonder if you don't simply need to sort the array here